### PR TITLE
Adding documentation action

### PR DESCRIPTION
**Pull request by jerinphilip**
Uploads latest documentation to repository's gh-pages branch.

There ~~can be~~ is some trickery to do this per version and also latest, like how docs are normally structured. Have latest `v<major>.<minor>.<suffix>` corresponding to releases as well.

Here's demonstration of a successful run on my fork:

* https://github.com/jerinphilip/bergamot-translator/runs/2625000572?check_suite_focus=true
* https://jerinphilip.github.io/bergamot-translator/ (sample-render)

Also could have doxygen warn about undocumented members, but maybe in a later PR.

The yaml is borrowed from @graemenail, upload-artifact ~~cut~~ only on pull request, gh-pages publish moved in.

### CODE REVIEW

**Comment by qianqianzhu** [comment]
There are two things:
 -  Can we only include *.h in generating Library API. There is a bug in Exhale if we both include *.h and *.cpp, the global functions/variables declared could appear in *.cpp.  You can see that this bug also happens in [byte_array_util.h](https://jerinphilip.github.io/bergamot-translator/api/file_src_translator_byte_array_util.h.html#file-src-translator-byte-array-util-h) and [byte_array_util.cpp](https://jerinphilip.github.io/bergamot-translator/api/file_src_translator_byte_array_util.cpp.html#file-src-translator-byte-array-util.cpp). This is super annoying and I haven't found a way to solve it for Marian docs (we have to include several .cpp files there). If you don't need *.cpp files, let's only include *.h in generating Library API. 
 - Can we host docs webpages in a more proper place? Like `bergamot-translator.github.io`. I don't think users will build the docs themselves locally. They just click the link.

**Comment by jerinphilip** [comment]
> Can we only include *.h in generating Library API. There is a bug in Exhale if we both include *.h and *.cpp, the global functions/variables declared could appear in *.cpp.

This is okay, right? From a user perspective? Are there any negatives to this? If and when marian has a workaround, we'll import it here? I can investigate this in a later PR if need be. I don't want to block this one on what could be a harder task (I don't know the impact of hiding *.cpp from Doxyfile, so..)

>  - Can we host docs webpages in a more proper place? Like `bergamot-translator.github.io`. I don't think users will build the docs themselves locally. They just click the link.

Is it not obvious that's where this is going? It pushes to gh-pages in this repository, which once this PR is in and gh-pages ready I will configure in settings to publish. The render is for demonstration on my repo, while I have push rights here I'm not overdoing it that's all.

**Comment by graemenail** [comment]
> Is it not obvious that's where this is going? It pushes to gh-pages in this repository, which once this PR is in and gh-pages ready I will configure in settings to publish. The render is for demonstration on my repo, while I have push rights here I'm not overdoing it that's all.

It should end up at https://browsermt.github.io/bergamot-translator

File: .github/workflows/doc.yml
Comment (jerinphilip) [reply .github/workflows/doc.yml]: Thanks, didn't know this! :D

**Comment by graemenail** [comment]
You may want to keep an artifact build for PRs as a check before merging

**Comment by jerinphilip** [comment]
Note that to not have dead links later and some clean structuring, we could attempt the version thing as follows. 

* https://browsermt.github.io/bergamot-translator/latest (built from main)
* https://browsermt.github.io/bergamot-translator/vM.m.p (built from tags)

Do we only care for the current documentation like https://marian-nmt.github.io/docs/api/?

**Comment by jerinphilip** [comment]
> You may want to keep an artifact build for PRs as a check before merging

I'll have to download and launch a server right? Does it render on the web? In this case won't pulling the PR/branch and running locally be more convenient?

**Comment by qianqianzhu** [comment]
> This is okay, right? From a user perspective? Are there any negatives to this?
Personally, I would expect API reference to only appears in header files. But anyway, it is not quite critical. You may leave it for a later PR.

**Comment by graemenail** [comment]
> > You may want to keep an artifact build for PRs as a check before merging
> 
> I'll have to download and launch a server right? Does it render on the web? If not won't pulling the PR/branch and running locally be more convenient?

Should be just HTML, so you can download the artifact and just open it in the browser.

File: .github/workflows/doc.yml
Comment (graemenail) [inline .github/workflows/doc.yml:42]:
```suggestion
        if: ${{ github.event_name == 'push'}}
        with:
```

File: .github/workflows/doc.yml
Comment (graemenail) [inline .github/workflows/doc.yml:50]:
```suggestion
        if: ${{ github.event_name == 'pull_request'}}
        with:
```

File: .github/workflows/doc.yml
Comment (graemenail) [inline .github/workflows/doc.yml:5]:
```suggestion
    branches: [ $default-branch, ci-sandbox ]
  pull_request:
    branches: [ $default-branch ]
```

**Review by graemenail — COMMENTED** [review]
This is more what I had in mind for artifacts and PRs

File: .github/workflows/doc.yml
Comment (jerinphilip) [reply .github/workflows/doc.yml:5]: I can remove this following a final approve, right now I demonstrate yaml works using ci-sandbox. Hope you understand.

What I'm trying for is to catch github.ref and push to target folder now attempting to strip `refs/tags`. It'll solve the issue with ci-sandbox.

https://github.community/t/how-to-get-just-the-tag-name/16241

Comment (graemenail) [reply .github/workflows/doc.yml:5]: This adds a PR trigger; I left the `ci-sandbox` as I assumed you were testing something.

**Comment by jerinphilip** [comment]
The behaviour is updated as follows. 

* https://browsermt.github.io/bergamot-translator/<branch> (built from branch)
* https://browsermt.github.io/bergamot-translator/vM.m.p (built from tags that map to semantic versions)

**Comment by jerinphilip** [comment]
After some offline inputs from @graemenail and @kpu, the strategy is now to keep the generated files in a docs repository (https://github.com/browsermt/docs). There's an SSH public-key (browsermt/docs) and private-key (here, bergamot-translator) to enable the same, stored in secrets.

ci-sandbox documentation is now live for a demo at: 
* https://browser.mt/docs/ci-sandbox/

There's a runtime patch to update `BERGAMOT_VERSION` file which sphinx uses to point to the branch in place.

**Review by graemenail — APPROVED** [review]
Looks great.

I suggest you add a few lines in the README on `main` over at [browsermt/docs](https://github.com/browsermt/docs) to explain that the content is located at `https://browser.mt/docs/` and is built from this repo.

**Review by qianqianzhu — APPROVED** [review]

**Pull request merged**
